R/expected_event.R

```diff
@@ -235,20 +223,32 @@ expected_event <- function(
     ans <- as.numeric(sum(df$nbar))
   } else {
     sf_start_fail <- stats::stepfun(start_fail, c(0, start_fail), right = FALSE)
-    ans <- df %>%
-      transmute(
-        t = end_fail,
-        fail_rate = fail_rate_var,
-        event = nbar,
-        start_fail = sf_start_fail(start_fail)
-      ) %>%
-      group_by(start_fail) %>%
-      summarize(
-        fail_rate = first(fail_rate),
-        event = sum(event)
-      ) %>%
-      mutate(t = start_fail) %>%
-      select(t, fail_rate, event)
+    ans <- data.frame(
+      t = df$end_fail,
+      fail_rate = df$fail_rate_var,
+      event = df$nbar,
+      start_fail = sf_start_fail(df$start_fail)
+    )
+    ans <- by(
+      ans, ans$start_fail,
+      function(data) {
+        data.frame(
+          start_fail = data$start_fail[1],
```

Comment:
Could you please remind me why we have `start_fail` in the `data.frame`?

Reply:
`start_fail` is later renamed to `t` and included in the output when `simple = FALSE`

Reply:
Please correct me if my understanding is wrong: it looks the `by()` function call is a base-R operation to replace `group_by()` in dplyr. Following the `group_by` function, we summarize 2 variables (`fail_rate` and `event`) in the old code. But in `by()`, we create 3 variables (`start_fail`, `fail_rate`, and `event`). Could you please help me understand the difference?

Reply:
I apologize for misunderstanding your question earlier. The discrepancy you are noticing is due to differences in behavior between the two approaches. `group_by()`/`summarize()` include the grouping variable in the final result. `by()` omits the grouping variable, so I had manually include it. Below is example code that demonstrates why `start_fail` needs to be defined in the function applied with `by()`. There is also a nice [dplyr vignette](https://cran.r-project.org/web/packages/dplyr/vignettes/base.html) that shows various options for replacing dplyr functions with base R functions.

```R
# original dplyr
library("dplyr")
ans %>%
  group_by(start_fail) %>%
  summarize(
    fail_rate = first(fail_rate),
    event = sum(event)
  )
## # A tibble: 3 × 3
##   start_fail fail_rate event
##        <dbl>     <dbl> <dbl>
## 1          0      0.05  2.91
## 2          1      0.02  1.11
## 3          2      0.01  3.45

# current replacement with by() + do.call(rbind)
current <- by(
  ans, ans$start_fail,
  function(data) {
    data.frame(
      start_fail = data$start_fail[1],
      fail_rate = data$fail_rate[1],
      event = sum(data$event)
    )
  }
)
do.call(rbind, current)
##   start_fail fail_rate    event
## 0          0      0.05 2.911773
## 1          1      0.02 1.113334
## 2          2      0.01 3.454813

# by() without including start_fail
without_start_fail <- by(
  ans, ans$start_fail,
  function(data) {
    data.frame(
      fail_rate = data$fail_rate[1],
      event = sum(data$event)
    )
  }
)
do.call(rbind, without_start_fail)
##   fail_rate    event
## 0      0.05 2.911773
## 1      0.02 1.113334
## 2      0.01 3.454813
```

Reply:
Thank you for your patient explanation. This lesson is quite valuable to learn